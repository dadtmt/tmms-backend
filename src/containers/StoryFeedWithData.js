import R from 'ramda'
import React, { Component } from 'react'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'

import StoryFeed from '../components/StoryFeed'

const getAllPagesQuery = gql`
  query GetAllPages($first: Int, $after: String, $orderBy: [PageOrderByArgs]) {
    viewer {
      allPages(first: $first, after: $after, orderBy: $orderBy) {
        edges {
          cursor
          node {
            id
            text
            choices {
              edges{
                node {
                  id
                  text
                  made
                }
              }
            }
          }
        }
      }
    }
  }
`

export const transformDataToPages = R.pipe(
  R.path(['viewer', 'allPages', 'edges']),
  R.map(R.pipe(
    R.prop('node'),
    R.over(
      R.lensProp('choices'),
      R.pipe(
        R.prop('edges'),
        R.map(R.prop('node'))
      )
    )
  ))
)

class Feed extends Component {

  constructor() {
    super()
    this.state = {
      pages: []
    }
  }

  componentWillReceiveProps(nextProps) {
    const { data } = nextProps
    if (!data.loading && !data.error) {
      this.setState({
        pages: transformDataToPages(data)
      })
    }
  }

  render() {
    return this.props.data.loading
      ? <div>Loading...</div>
      : <StoryFeed pages={this.state.pages} />
  }
}

Feed.propTypes = {
  data: React.PropTypes.object.isRequired
}

export default graphql(getAllPagesQuery)(Feed)
