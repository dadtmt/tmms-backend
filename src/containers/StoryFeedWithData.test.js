import { transformDataToPages } from './StoryFeedWithData'

describe('transformDataToPages', () => {
  it('return pages array', () => {
    const data = {
      'viewer': {
        'allPages': {
          'edges': [
            {
              'cursor': 'Y3Vyc29yOjU0',
              'node': {
                'id': 'UGFnZTo1NA==',
                'text': '{\'entityMap\':{},\'blocks\':[{\'key\':\'1q0f0\',\'text\':\'aa\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}',
                'choices': {
                  'edges': [
                    {
                      'node': {
                        'id': 'Q2hvaWNlOjM3',
                        'text': '{\'entityMap\':{},\'blocks\':[{\'key\':\'bo88m\',\'text\':\'a\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}',
                        'made': false
                      }
                    }
                  ]
                }
              }
            },
            {
              'cursor': 'Y3Vyc29yOjUz',
              'node': {
                'id': 'UGFnZTo1Mw==',
                'text': '{\'entityMap\':{},\'blocks\':[{\'key\':\'a1mjs\',\'text\':\'antoher page\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}',
                'choices': {
                  'edges': [
                    {
                      'node': {
                        'id': 'Q2hvaWNlOjM2',
                        'text': '{\'entityMap\':{},\'blocks\':[{\'key\':\'4cmds\',\'text\':\'rety\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}',
                        'made': true
                      }
                    }
                  ]
                }
              }
            },
            {
              'cursor': 'Y3Vyc29yOjUy',
              'node': {
                'id': 'UGFnZTo1Mg==',
                'text': '{\'entityMap\':{},\'blocks\':[{\'key\':\'afcg\',\'text\':\'first PAge\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}',
                'choices': {
                  'edges': [
                    {
                      'node': {
                        'id': 'Q2hvaWNlOjM1',
                        'text': '{\'entityMap\':{},\'blocks\':[{\'key\':\'6reaa\',\'text\':\'2nd choice\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}',
                        'made': false
                      }
                    },
                    {
                      'node': {
                        'id': 'Q2hvaWNlOjM0',
                        'text': '{\'entityMap\':{},\'blocks\':[{\'key\':\'6reaa\',\'text\':\'ffff\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}',
                        'made': false
                      }
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
    expect(transformDataToPages(data)).toMatchSnapshot()
  })
})
