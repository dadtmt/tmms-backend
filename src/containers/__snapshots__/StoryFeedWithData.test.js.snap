exports[`transformDataToPages return pages array 1`] = `
Array [
  Object {
    "choices": Array [
      Object {
        "id": "Q2hvaWNlOjM3",
        "made": false,
        "text": "{\'entityMap\':{},\'blocks\':[{\'key\':\'bo88m\',\'text\':\'a\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}",
      },
    ],
    "id": "UGFnZTo1NA==",
    "text": "{\'entityMap\':{},\'blocks\':[{\'key\':\'1q0f0\',\'text\':\'aa\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}",
  },
  Object {
    "choices": Array [
      Object {
        "id": "Q2hvaWNlOjM2",
        "made": true,
        "text": "{\'entityMap\':{},\'blocks\':[{\'key\':\'4cmds\',\'text\':\'rety\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}",
      },
    ],
    "id": "UGFnZTo1Mw==",
    "text": "{\'entityMap\':{},\'blocks\':[{\'key\':\'a1mjs\',\'text\':\'antoher page\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}",
  },
  Object {
    "choices": Array [
      Object {
        "id": "Q2hvaWNlOjM1",
        "made": false,
        "text": "{\'entityMap\':{},\'blocks\':[{\'key\':\'6reaa\',\'text\':\'2nd choice\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}",
      },
      Object {
        "id": "Q2hvaWNlOjM0",
        "made": false,
        "text": "{\'entityMap\':{},\'blocks\':[{\'key\':\'6reaa\',\'text\':\'ffff\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}",
      },
    ],
    "id": "UGFnZTo1Mg==",
    "text": "{\'entityMap\':{},\'blocks\':[{\'key\':\'afcg\',\'text\':\'first PAge\',\'type\':\'unstyled\',\'depth\':0,\'inlineStyleRanges\':[],\'entityRanges\':[],\'data\':{}}]}",
  },
]
`;
